Better Octane support

luischavez opened the pull request:
Hi, everyone I'm very excited with octane so I'm trying to get a stable release of Livewire that works perfectly with octane, this MR contains a few changes to solve some of the common octane problem that I found, here you have a brief description:

1. Fixed properties and computed properties in full page component that preserves along requests (Moved to #3947).
2. Fixed __invoke Route parameter binding (Moved to #3946).
3. Fixed memory issues with some of the clases in Features folder (Moved to #3940).

### Fixed properties and computed properties in full page component that preserves along requests.
The Illuminate/Routing/Route keeps in memory an instance of the component:
[https://github.com/laravel/framework/blob/8.x/src/Illuminate/Routing/Route.php#L271](https://github.com/laravel/framework/blob/8.x/src/Illuminate/Routing/Route.php#L271)
```PHP
  /**
     * Get the controller instance for the route.
     *
     * @return mixed
     */
    public function getController()
    {
        if (! $this->controller) {
            $class = $this->parseControllerCallback()[0];

            $this->controller = $this->container->make(ltrim($class, '\\'));
        }

        return $this->controller;
    }
```
so we need to create always a fresh instance of the component on the __invoke method.

### Fixed __invoke Route parameter binding.
This was really hard to found and to be honest I really don't know why is this causing the issue, but if you not set the container in the route created in the method makeRequestFromUrlAndMethod then the routing system can't resolve any Route parameter binding anymore, so after set the container in the route all works good.
```PHP
    protected function makeRequestFromUrlAndMethod($url, $method = 'GET')
    {
        $request = Request::create($url, $method);

        if ($session = request()->getSession()) {
            $request->setLaravelSession($session);
        }

        $request->setUserResolver(request()->getUserResolver());

        $route = app('router')->getRoutes()->match($request);

        // For some reason without this octane breaks the route parameter binding.
        $route->setContainer(app());

        $request->setRouteResolver(function () use ($route) {
            return $route;
        });

        return $request;
    }
```

### Fixed memory issues with some of the clases in Features folder.
We need to be careful when working with octane and Livewire::listen, because Livewire instance is a singleton it will keep in memory any reference inside the listener, for example in SupportComponentTraits:
```PHP
    protected $componentIdMethodMap = [];

    function __construct()
    {
        Livewire::listen('component.hydrate', function ($component) {
            $component->initializeTraits();

            foreach (class_uses_recursive($component) as $trait) {
                $hooks = [
                    'hydrate',
                    'mount',
                    'updating',
                    'updated',
                    'rendering',
                    'rendered',
                    'dehydrate',
                ];

                foreach ($hooks as $hook) {
                    $method = $hook.class_basename($trait);

                    if (method_exists($component, $method)) {
                        $this->componentIdMethodMap[$component->id][$hook][] = [$component, $method];
                    }
                }
            }

            $methods = $this->componentIdMethodMap[$component->id]['hydrate'] ?? [];

            foreach ($methods as $method) {
                ImplicitlyBoundMethod::call(app(), $method);
            }
        });
```
the $componentIdMethodMap variable referenced in the listener will be the same on each request, so the listener will add more and more callables to the variable and it grows without limit, my solution is keep all the "state" in the component so I added a few methods on the component to store the state. 

@calebporzio could you take a look on this?

luischavez committed:
Improved octane support

luischavez committed:
Fixed mount listener on SupportComponentTraits

luischavez committed:
Added component state test

luischavez committed:
Reverted changes in component __invoke method signature

luischavez committed:
Fixed octane route binding on component __invoke

luischavez committed:
Fixed octane route binding on component __invoke

luischavez commented:
Related issues:
#3495 
#3899
#2878
#2720

calebporzio commented:
Thanks for submitting this @luischavez, it's a huge help!

Is it possible to separate out these changes into separate PRs?

My reasoning is this:
- I'd like to merge the stuff with the `$route->setContainer` and `__invoke()` changes right away.
- I think the `->getState()` strategy is good and their should be a mechanism like that, however, I'm not sure about putting it on the component class itself. Sometimes we need cross-component data when components render nested components. So separating it out into a different PR allows us to talk more about that solution specifically.
- @joshhanley has some ideas about the "Fixed memory issues" part that I think he wants to PR to Octane itself.
- It looks like there might be a bug with you removing the array of hooks whitelist entirely (`update`, `hydrate`, `mount` etc...)

Thanks again, but I'm not comfortable merging this all at once.

luischavez committed:
Reverted memory fixes

luischavez committed:
Removed Component state tests

luischavez commented:
@calebporzio no problem, I put the changes related to the state into this PR: #3940

luischavez committed:
Reverted to original changes

luischavez committed:
Clean PR

luischavez commented:
@joshhanley Thanks, as you suggest I split the changes on 3 PR:
#3947 
#3946 
#3940 

Also I reverted the changes on this PR as I original submit.

joshhanley commented:
@luischavez great, thanks! 🙂

calebporzio commented:
Thanks @luischavez - I'm closing this one
